### PR TITLE
Refactor lora

### PR DESCRIPTION
**Pull request by grimoire**
s-lora is hard to maintain.

### CODE REVIEW

**Comment by lvhan028** [comment]
Should we update `api_server_lora.md`?

File: lmdeploy/pytorch/kernels/cuda/fused_lora.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+               max_rank: int, max_seqlen: int):
+    """fused lora."""
+
+    def grid(META):
+        ret = ((triton.cdiv(max_seqlen, META['BLOCK_SIZE_M']) *
+                triton.cdiv(N, META['BLOCK_SIZE_N'])), batch_size)
+        return ret
+
+    assert input.dim() == 2
+    batch_size = seq_lens.numel()
```
Comment (AllentDan) [inline lmdeploy/pytorch/kernels/cuda/fused_lora.py:157]: Here, we know how many sessions are running. Why do we use `seq_lens.numel()`?

**Comment by AllentDan** [comment]
May update `pipeline.html` as well.

File: lmdeploy/pytorch/kernels/cuda/fused_lora.py (hunk 1 above)
Comment (grimoire) [reply lmdeploy/pytorch/kernels/cuda/fused_lora.py:157]: cudagraph support require static grid size.

**Comment by irexyc** [comment]
https://github.com/InternLM/lmdeploy/blob/85cf6e3f3f780e83f33cda608f5e312c5502cda2/docs/zh_cn/llm/pipeline.md

Do `adapters` only support local paths?

```
  File "/home/chenxin/ws3/topk/lmdeploy/pytorch/engine/model_agent.py", line 248, in _build_model
    add_adapters(patched_model,
  File "/home/chenxin/miniconda3/envs/38/lib/python3.8/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/chenxin/ws3/topk/lmdeploy/pytorch/models/patch.py", line 298, in add_adapters
    state_dict = torch.load(checkpoint_path, map_location=device)
  File "/home/chenxin/miniconda3/envs/38/lib/python3.8/site-packages/torch/serialization.py", line 997, in load
    with _open_file_like(f, 'rb') as opened_file:
  File "/home/chenxin/miniconda3/envs/38/lib/python3.8/site-packages/torch/serialization.py", line 444, in _open_file_like
    return _open_file(name_or_buffer, mode)
  File "/home/chenxin/miniconda3/envs/38/lib/python3.8/site-packages/torch/serialization.py", line 425, in __init__
    super().__init__(open(name, mode))
FileNotFoundError: [Errno 2] No such file or directory: 'chenchi/lora-chatglm2-6b-guodegang/adapter_model.bin'
```

**Comment by grimoire** [comment]
> Do `adapters` only support local paths?

Add download adapters.

**Comment by irexyc** [comment]
With `do_sample=False` and same input, the output of the first request is different from the output of the subsequent request. Is this as expected?

```python
from lmdeploy import pipeline, GenerationConfig, PytorchEngineConfig
backend_config = PytorchEngineConfig(session_len=2048,
                                     adapters=dict(lora_name_1='chenchi/lora-chatglm2-6b-guodegang'))
pipe = pipeline('/mnt/140/chatglm2-6b/',
                log_level='INFO',
                backend_config=backend_config)
prompts = [[{
    'role': 'user',
    'content': '您猜怎么着'
}]]
response = pipe(prompts, adapter_name='lora_name_1')
print(response[0].text)
response = pipe(prompts, adapter_name='lora_name_1')
print(response[0].text)
response = pipe(prompts, adapter_name='lora_name_1')
print(response[0].text)

#  我不知道您想说什么,可以请您把您的想法、问题、观点、建议或者疑问提出来,我会尽力帮助您解答。
#  没有啊，您给我说。
#  没有啊，您给我说。
```

**Comment by grimoire** [comment]
> With `do_sample=False` and same input, the output of the first request is different from the output of the subsequent request. Is this as expected?

Fixed.

**Review by AllentDan — APPROVED** [review]
LGTM

**Pull request merged**